pkg/tiflash/client.go

```diff
@@ -54,6 +54,11 @@ func (c Client) WithTimeout(timeout time.Duration) *Client {
 	return &c
 }
 
+func (c Client) WithBeforeRequest(callback func(req *http.Request)) *Client {
+	c.httpClient.BeforeRequest = callback
+	return &c
+}
+
```

Comment:
There's an incorrect impl at pd/client.go. It will set `BeforeRequest` to the base http client reference.

We can fix like [this](https://github.com/pingcap/tidb-dashboard/pull/1050/files#diff-45e2e42c28256d3b1f144bc786d0e72deade31665db3a188aa9b32fe516e583eR86-R89). And invoke `c.httpClient.WithBeforeRequest` at L64.

Reply:
Sorry, this fix still has the problem of shared pointers.

You might be blocked by this pr. https://github.com/pingcap/tidb-dashboard/pull/1069

Reply:
Got it!